storm-core/src/native/worker-launcher/impl/worker-launcher.c

```diff
@@ -414,33 +414,30 @@ static int copy_file(int input, const char* in_filename,
   return 0;
 }
 
-int setup_stormdist(FTSENT* entry, uid_t euser) {
+static int setup_permissions(FTSENT* entry, uid_t euser, int user_write) {
   if (lchown(entry->fts_path, euser, launcher_gid) != 0) {
-    fprintf(ERRORFILE, "Failure to exec app initialization process - %s\n",
-      strerror(errno));
+    fprintf(ERRORFILE, "Failure to exec app initialization process - %s, fts_path=%s\n",
+            strerror(errno), entry->fts_path);
      return -1;
   }
   mode_t mode = entry->fts_statp->st_mode;
-  mode_t new_mode = (mode & (S_IRWXU)) | S_IRGRP | S_IWGRP;
-  if ((mode & S_IXUSR) == S_IXUSR) {
-    new_mode = new_mode | S_IXGRP;
+  mode_t new_mode = (mode & (S_IRUSR | S_IXUSR)) | S_IRGRP | S_IWGRP;
+  if (user_write) {
+    new_mode = new_mode | S_IWUSR;
```

Comment:
It was not completely obvious from reading the code what was happening here.  Could we put in some comments at a high level how the permissions are setup and why they are setup that way.